src/bsmHandler.cpp

```diff
@@ -222,58 +226,76 @@ bool BSMHandler::process( const std::string& message_json ) {
 
         rapidjson::Value& data = payload["data"];
 
-        if (!data.HasMember("coreData")) {
+        if (!data.HasMember("value")) {
+            result_ = ResultStatus::MISSING;
+
+            return false;
+        }
+
+        rapidjson::Value& value = data["value"];
+
+        if (!value.HasMember("BasicSafetyMessage")) {
             result_ = ResultStatus::MISSING;
 
             return false;
         }
 
-        rapidjson::Value& core_data = data["coreData"];
+        rapidjson::Value& basicSafetyMessage = value["BasicSafetyMessage"];
+
+        if (!basicSafetyMessage.HasMember("coreData")) {
+            result_ = ResultStatus::MISSING;
+
+            return false;
+        }
+
+        rapidjson::Value& core_data = basicSafetyMessage["coreData"];
 
         if (!core_data.HasMember("speed")) {
             result_ = ResultStatus::MISSING;
 
             return false;
         }
         
-        if (!core_data["speed"].IsDouble()) {
+        if (!core_data["speed"].IsInt()) {
             result_ = ResultStatus::OTHER;
 
             return false;
         }
 
-        speed = core_data["speed"].GetDouble();
-        bsm_.set_velocity(speed);
+        // Only set speed if it's not the J2735 "unavailable" value
+        if (core_data["speed"].GetInt() != J2735_SPEED_UNAVAILABLE) {
+            speed = core_data["speed"].GetInt() * 0.02;
+            bsm_.set_velocity(speed);
+        }
 
         if (is_active<kVelocityFilterFlag>() && vf_.suppress(speed)) {
             result_ = ResultStatus::SPEED;
         }
 
-        if (!core_data.HasMember("position")) {
+        // Check if position data is available
+        if (!core_data.HasMember("lat") || !core_data.HasMember("long")) {
             result_ = ResultStatus::MISSING;
 
             return false;
         }
 
-        rapidjson::Value& position = core_data["position"];
-
-        if (!position.HasMember("latitude") || !position.HasMember("longitude")) {
-            result_ = ResultStatus::MISSING;
+        if (!core_data["lat"].IsInt() || !core_data["long"].IsInt()) {
+            result_ = ResultStatus::OTHER;
 
             return false;
         }
 
-        if (!position["latitude"].IsDouble() || !position["longitude"].IsDouble()) {
-            result_ = ResultStatus::OTHER;
-
-            return false;
+        // Only set latitude if it's not the J2735 "unavailable" value
+        if (core_data["lat"].GetInt() != J2735_LATITUDE_UNAVAILABLE) {
+            latitude = core_data["lat"].GetInt() * 1e-7;
+            bsm_.set_latitude(latitude);
         }
-        
-        latitude = position["latitude"].GetDouble();
-        longitude = position["longitude"].GetDouble();
 
-        bsm_.set_latitude(latitude); 
-        bsm_.set_longitude(longitude); 
+        // Only set longitude if it's not the J2735 "unavailable" value
+        if (core_data["long"].GetInt() != J2735_LONGITUDE_UNAVAILABLE) {
+            longitude = core_data["long"].GetInt() * 1e-7;
```

Comment:
suggestion: add comment explaining the math

Reply:
Addressed.